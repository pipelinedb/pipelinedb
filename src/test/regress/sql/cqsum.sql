SET debug_sync_stream_insert = 'on';

-------------------------------------------------------------------------------
-- Integer sums
CREATE CONTINUOUS VIEW test_int8 AS SELECT k::text, SUM(v::int8) FROM int_stream GROUP BY k;
CREATE CONTINUOUS VIEW test_int4 AS SELECT k::text, SUM(v::int4) FROM int_stream GROUP BY k;
CREATE CONTINUOUS VIEW test_int2 AS SELECT k::text, SUM(v::int2) FROM int_stream GROUP BY k;

ACTIVATE test_int8, test_int4, test_int2;

INSERT INTO int_stream (k, v) VALUES ('x', 10), ('x', 10), ('x', 10);
INSERT INTO int_stream (k, v) VALUES ('y', 10), ('y', 10), ('y', 10);

DEACTIVATE;

SELECT * FROM test_int8;
SELECT * FROM test_int4;
SELECT * FROM test_int2;

ACTIVATE test_int8, test_int4, test_int2;

INSERT INTO int_stream (k, v) VALUES ('x', 1), ('x', 2), ('x', 3);
INSERT INTO int_stream (k, v) VALUES ('y', 10), ('y', 10), ('z', 10);

DEACTIVATE;

SELECT * FROM test_int8;
SELECT * FROM test_int4;
SELECT * FROM test_int2;

-------------------------------------------------------------------------------
-- Float sums
CREATE CONTINUOUS VIEW test_float8 AS SELECT k::text, SUM(v::float8) FROM float_stream GROUP BY k;
CREATE CONTINUOUS VIEW test_float4 AS SELECT k::text, SUM(v::float4) FROM float_stream GROUP BY k;

ACTIVATE test_float8, test_float4;

INSERT INTO float_stream (k, v) VALUES ('x', 10.2), ('x', 1.2), ('x', 0.000001);
INSERT INTO float_stream (k, v) VALUES ('y', -10.3), ('y', 1.2e6), ('y', '100.4');

DEACTIVATE;

SELECT * FROM test_float8_pdb;
SELECT * FROM test_float4_pdb;

ACTIVATE test_float8, test_float4;

INSERT INTO float_stream (k, v) VALUES ('x', 10.2), ('x', 1.2), ('x', 1.2e-4);
INSERT INTO float_stream (k, v) VALUES ('y', -10.3), ('y', 1.2e6), ('y', '100.4');

DEACTIVATE;

SELECT * FROM test_float8;
SELECT * FROM test_float4;

-------------------------------------------------------------------------------
-- Cash sums
CREATE CONTINUOUS VIEW test_cash AS SELECT k::text, SUM(v::money) FROM cash_stream GROUP BY k;

ACTIVATE test_cash;

INSERT INTO cash_stream (k, v) VALUES ('x', 10.2), ('x', 1.2), ('x', 0.10);
INSERT INTO cash_stream (k, v) VALUES ('y', -10), ('y', 10), ('y', 0);

DEACTIVATE;

SELECT * FROM test_cash;

ACTIVATE test_cash;

INSERT INTO cash_stream (k, v) VALUES ('x', '500.50');
INSERT INTO cash_stream (k, v) VALUES ('y', '0.01');

DEACTIVATE;

SELECT * FROM test_cash;

-------------------------------------------------------------------------------
-- Numeric sums
CREATE CONTINUOUS VIEW test_numeric AS SELECT k::text, SUM(v::numeric) FROM numeric_stream GROUP BY k;

ACTIVATE test_numeric;

INSERT INTO numeric_stream (k, v) VALUES ('x', 10.0002), ('x', 0.0001), ('x', -0.10);
INSERT INTO numeric_stream (k, v) VALUES ('y', 1.004e5), ('y', 0.4), ('y', 0);

DEACTIVATE;

SELECT * FROM test_numeric;

ACTIVATE test_numeric;

INSERT INTO numeric_stream (k, v) VALUES ('x', '10000000000000000000000000000000000000000000000000000');
INSERT INTO numeric_stream (k, v) VALUES ('y', '-0000000000000000000000000000000000000.00000000000000000000000000000000001');

DEACTIVATE;

SELECT * FROM test_numeric;

-------------------------------------------------------------------------------
-- Interval sum
CREATE CONTINUOUS VIEW test_interval AS SELECT k::text, SUM(ts1::timestamp - ts0::timestamp) FROM interval_stream GROUP BY k;

ACTIVATE test_interval;

INSERT INTO interval_stream (k, ts0, ts1) VALUES ('x', '2014-01-01', '2014-01-02'), ('x', '2014-01-01', '2014-02-01');
INSERT INTO interval_stream (k, ts0, ts1) VALUES ('x', '2014-01-01', '2014-01-02'), ('x', '2014-01-01', '2014-02-01');

DEACTIVATE;

SELECT * FROM test_interval;

ACTIVATE test_interval;

INSERT INTO interval_stream (k, ts0, ts1) VALUES ('x', '2014-12-31', '2015-01-01'), ('x', '2015-02-28', '2015-03-01');
INSERT INTO interval_stream (k, ts0, ts1) VALUES ('y', '2014-01-01', '2014-01-02'), ('z', '2014-01-01', '2014-02-01');

DEACTIVATE;

SELECT * FROM test_interval;

