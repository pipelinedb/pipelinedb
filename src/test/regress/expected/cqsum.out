SET debug_sync_stream_insert = 'on';
-------------------------------------------------------------------------------
-- Integer sums
CREATE CONTINUOUS VIEW test_int8 AS SELECT k::text, SUM(v::int8) FROM int_stream GROUP BY k;
CREATE CONTINUOUS VIEW test_int4 AS SELECT k::text, SUM(v::int4) FROM int_stream GROUP BY k;
CREATE CONTINUOUS VIEW test_int2 AS SELECT k::text, SUM(v::int2) FROM int_stream GROUP BY k;
ACTIVATE test_int8, test_int4, test_int2;
INSERT INTO int_stream (k, v) VALUES ('x', 10), ('x', 10), ('x', 10);
INSERT INTO int_stream (k, v) VALUES ('y', 10), ('y', 10), ('y', 10);
DEACTIVATE;
SELECT * FROM test_int8;
 k | sum 
---+-----
 x |  30
 y |  30
(2 rows)

SELECT * FROM test_int4;
 k | sum 
---+-----
 x |  30
 y |  30
(2 rows)

SELECT * FROM test_int2;
 k | sum 
---+-----
 x |  30
 y |  30
(2 rows)

ACTIVATE test_int8, test_int4, test_int2;
INSERT INTO int_stream (k, v) VALUES ('x', 1), ('x', 2), ('x', 3);
INSERT INTO int_stream (k, v) VALUES ('y', 10), ('y', 10), ('z', 10);
DEACTIVATE;
SELECT * FROM test_int8;
 k | sum 
---+-----
 x |  36
 y |  50
 z |  10
(3 rows)

SELECT * FROM test_int4;
 k | sum 
---+-----
 x |  36
 y |  50
 z |  10
(3 rows)

SELECT * FROM test_int2;
 k | sum 
---+-----
 x |  36
 y |  50
 z |  10
(3 rows)

-------------------------------------------------------------------------------
-- Float sums
CREATE CONTINUOUS VIEW test_float8 AS SELECT k::text, SUM(v::float8) FROM float_stream GROUP BY k;
CREATE CONTINUOUS VIEW test_float4 AS SELECT k::text, SUM(v::float4) FROM float_stream GROUP BY k;
ACTIVATE test_float8, test_float4;
INSERT INTO float_stream (k, v) VALUES ('x', 10.2), ('x', 1.2), ('x', 0.000001);
INSERT INTO float_stream (k, v) VALUES ('y', -10.3), ('y', 1.2e6), ('y', '100.4');
DEACTIVATE;
SELECT * FROM test_float8_pdb;
 k |    sum    
---+-----------
 x | 11.400001
 y | 1200090.1
(2 rows)

SELECT * FROM test_float4_pdb;
 k |     sum     
---+-------------
 x |        11.4
 y | 1.20009e+06
(2 rows)

ACTIVATE test_float8, test_float4;
INSERT INTO float_stream (k, v) VALUES ('x', 10.2), ('x', 1.2), ('x', 1.2e-4);
INSERT INTO float_stream (k, v) VALUES ('y', -10.3), ('y', 1.2e6), ('y', '100.4');
DEACTIVATE;
SELECT * FROM test_float8;
 k |    sum    
---+-----------
 x | 22.800121
 y | 2400180.2
(2 rows)

SELECT * FROM test_float4;
 k |     sum     
---+-------------
 x |     22.8001
 y | 2.40018e+06
(2 rows)

-------------------------------------------------------------------------------
-- Cash sums
CREATE CONTINUOUS VIEW test_cash AS SELECT k::text, SUM(v::money) FROM cash_stream GROUP BY k;
ACTIVATE test_cash;
INSERT INTO cash_stream (k, v) VALUES ('x', 10.2), ('x', 1.2), ('x', 0.10);
INSERT INTO cash_stream (k, v) VALUES ('y', -10), ('y', 10), ('y', 0);
DEACTIVATE;
SELECT * FROM test_cash;
 k |  sum   
---+--------
 x | $11.50
 y |  $0.00
(2 rows)

ACTIVATE test_cash;
INSERT INTO cash_stream (k, v) VALUES ('x', '500.50');
INSERT INTO cash_stream (k, v) VALUES ('y', '0.01');
DEACTIVATE;
SELECT * FROM test_cash;
 k |   sum   
---+---------
 x | $512.00
 y |   $0.01
(2 rows)

-------------------------------------------------------------------------------
-- Numeric sums
CREATE CONTINUOUS VIEW test_numeric AS SELECT k::text, SUM(v::numeric) FROM numeric_stream GROUP BY k;
ACTIVATE test_numeric;
INSERT INTO numeric_stream (k, v) VALUES ('x', 10.0002), ('x', 0.0001), ('x', -0.10);
INSERT INTO numeric_stream (k, v) VALUES ('y', 1.004e5), ('y', 0.4), ('y', 0);
DEACTIVATE;
SELECT * FROM test_numeric;
 k |   sum    
---+----------
 x |   9.9003
 y | 100400.4
(2 rows)

ACTIVATE test_numeric;
INSERT INTO numeric_stream (k, v) VALUES ('x', '10000000000000000000000000000000000000000000000000000');
INSERT INTO numeric_stream (k, v) VALUES ('y', '-0000000000000000000000000000000000000.00000000000000000000000000000000001');
DEACTIVATE;
SELECT * FROM test_numeric;
 k |                            sum                             
---+------------------------------------------------------------
 x | 10000000000000000000000000000000000000000000000000009.9003
 y |                 100400.39999999999999999999999999999999999
(2 rows)

-------------------------------------------------------------------------------
-- Interval sum
CREATE CONTINUOUS VIEW test_interval AS SELECT k::text, SUM(ts1::timestamp - ts0::timestamp) FROM interval_stream GROUP BY k;
ACTIVATE test_interval;
INSERT INTO interval_stream (k, ts0, ts1) VALUES ('x', '2014-01-01', '2014-01-02'), ('x', '2014-01-01', '2014-02-01');
INSERT INTO interval_stream (k, ts0, ts1) VALUES ('x', '2014-01-01', '2014-01-02'), ('x', '2014-01-01', '2014-02-01');
DEACTIVATE;
SELECT * FROM test_interval;
 k |    sum    
---+-----------
 x | @ 64 days
(1 row)

ACTIVATE test_interval;
INSERT INTO interval_stream (k, ts0, ts1) VALUES ('x', '2014-12-31', '2015-01-01'), ('x', '2015-02-28', '2015-03-01');
INSERT INTO interval_stream (k, ts0, ts1) VALUES ('y', '2014-01-01', '2014-01-02'), ('z', '2014-01-01', '2014-02-01');
DEACTIVATE;
SELECT * FROM test_interval;
 k |    sum    
---+-----------
 x | @ 66 days
 y | @ 1 day
 z | @ 31 days
(3 rows)

